File: fopa_sop_apk/lib/features/feat_user/domaines/dtos/create_user_dto.dart
import 'package:fopa_sop_apk/features/feat_auth/datas/models/user_model.dart';

class CreateUserDto {
  final String? email;
  final String name;
  final String phone;
  final String username;
  final String password;
  final UserRole role;

  CreateUserDto({
    this.email,
    required this.name,
    required this.phone,
    required this.username,
    required this.password,
    required this.role,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'name': name,
      'phone': phone,
      'username': username,
      'password': password,
      'role': role.name,
    };
    if (email != null && email!.isNotEmpty) map['email'] = email;
    return map;
  }
}

File: src/application/dtos/user/create-user.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { IsEmail, IsEnum, IsNotEmpty, IsOptional, IsString, MinLength } from 'class-validator';
import { UserRole } from '../../../domain/entities/user.entity';

export class CreateUserDto {
  @ApiPropertyOptional({ example: 'user@example.com' })
  @IsOptional()
  @IsEmail({}, { message: "L'email doit être valide" })
  email?: string;

  @ApiProperty({ example: 'Jane Doe' })
  @IsNotEmpty({ message: 'Le nom est requis' })
  @IsString()
  name: string;

  @ApiProperty({ example: '+2250123456789' })
  @IsNotEmpty({ message: 'Le téléphone est requis' })
  @IsString()
  phone: string;

  @ApiProperty({ example: 'jane_doe' })
  @IsNotEmpty({ message: "Le nom d'utilisateur est requis" })
  @IsString()
  username: string;

  @ApiProperty({ example: 'password123', minLength: 6 })
  @IsNotEmpty({ message: 'Le mot de passe est requis' })
  @IsString()
  @MinLength(6, { message: 'Le mot de passe doit contenir au moins 6 caractères' })
  password: string;

  @ApiPropertyOptional({ enum: UserRole, example: UserRole.CAISSIERE })
  @IsOptional({ message: 'Le rôle est requis' })
  @IsEnum(UserRole, { message: 'Le rôle doit être ADMIN, CAISSIERE ou CLIENT' })
  role?: UserRole | undefined = UserRole.CLIENT;

  constructor(name: string, phone: string, username: string, password: string, role: UserRole, email?: string) {
    this.email = email;
    this.name = name;
    this.phone = phone;
    this.username = username;
    this.password = password;
    this.role = role;
  }
}
